fix: Put app to the operation queue after refresh queue processing to avoid race condition [#18500]

andrii-korotkov-verkada opened the pull request:
Helps with #18500 

Adding app to the operation queue and refresh queue could cause waiting for resource for minutes to tens of minutes. Sync state operates on resources gathered from reconciliation, so if the app operation event is processed before the refresh one (when triggered on resource update/creation), the refresh doesn’t help sync to progress and it essentially needs to wait for another app refresh.

The fix seems to be to schedule app operation event after refresh event is finished processing. There’s one place where operation event is scheduled without refresh event (which can be kept there), and one place where refresh even is scheduled without the operation one during the app deletion handling https://github.com/argoproj/argo-cd/blob/3e2cfb138795e24df98c9745d813e4b7f1720dbd/controller/appcontroller.go#L2177. It’s probably safe to schedule operation even after that, since it has some code to check that app was deleted. If not, an update can be made to have refresh queue storing a tuple with app key and bool whether to enqueue app operation.

If there are issues: try keeping both old places to add to app operation queue and new addition after refresh.

Note on cherry pick: add to as many releases as you can. This can be a significant performance boost.

<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

andrii-korotkov-verkada commented:
I've deployed changes in a live environment and found that config map is not being waited at all - it's created and sync is progressing fast to next steps!

Note, there's been some network errors, like `lookup argocd-repo-server on <ip>:53: no such host`, but that's happening even on v2.11.3 and probably related to some network issues in our cluster.

agaudreault commented on controller/appcontroller.go:
```diff
@@ -2125,7 +2126,6 @@ func (ctrl *ApplicationController) newApplicationInformerAndLister() (cache.Shar
 				key, err := cache.MetaNamespaceKeyFunc(obj)
 				if err == nil {
 					ctrl.appRefreshQueue.AddRateLimited(key)
-					ctrl.appOperationQueue.AddRateLimited(key)
```
I think on Application Add/Update/Delete, we should add to the operation queue as soon as possible, to have the sync operation up to date. Refresh/Sync may take a few seconds, so it is nice to have the value reflected as early as possible. 

It does make sense to add to the operation queue when a refresh is completed IMO since it needs the latest status of the refresh (and sometimes the refresh is the one that starts the operation with auto-sync).

andrii-korotkov replied:
My understanding is that with refresh and operation scheduled at the same time it could always make sense to wait for refresh to complete, since otherwise operation can happen on a stale data. In one case on update when refresh and operation were scheduled with a delay and another immediate operation scheduling was present, I’ve kept the later one.

andrii-korotkov-verkada replied:
I'm putting it back, would update the PR shortly.

agaudreault reviewed (commented):
Change looks good to me. It makes sense to have the "operation" updated after the refresh. On a performance level, I dont think it will change much the quantity of refresh/operation items processed.

I think the current code relied on app refreshes being triggered very often and with optimizations in the latest version, these kinds timing problems reveal themselves!

andrii-korotkov commented:
> Change looks good to me. It makes sense to have the "operation" updated after the refresh. On a performance level, I dont think it will change much the quantity of refresh/operation items processed.
> 
> I think the current code relied on app refreshes being triggered very often and with optimizations in the latest version, these kinds timing problems reveal themselves!

The current code seems to need both app refresh and app operation to update a lot, at least the later one. I think in the scenario I’ve encountered the config map is often the only new thing created/updated in the corresponding sync wave, so there might be not much going on to trigger another app operation. One more note is that we ignore resource updates on status field change, which might contribute, though config map itself doesn’t have a status.

P.S. I’m writing from a personal account, currently left my work laptop at home.

andrii-korotkov-verkada commented:
I've still encountered one case when it waited for ~5 minutes on a healthy state of a config map. Though in many cases it waits at most ~1 minute.

andrii-korotkov-verkada commented:
This might be happening due to deduplication by queues if multiple refreshes are triggered in a quick succession. Maybe I can fix a part of it by adding a delay between refresh and app operation enqueuing, but not sure what to do with multiple refreshes. I'm not sure if the deduplication would happen if task done wasn't called yet but the processing has already started.

andrii-korotkov-verkada commented:
I've noticed that app refresh requests come in batches spread out by ~5-10ms or so. I'm thinking of adding a delay before adding to refresh queue to avoid a situation when first refresh picks up partial changes and next refreshes are getting deduplicated while the first one is still processing. Not 100% that's how dedup works, but assuming the worst case. I'll test with a delay.

andrii-korotkov-verkada commented:
I don't see an entry in the logs for `Requesting app refresh caused by object update` when ConfigMap is created though, only several minutes later sometimes, which is concerning.

andrii-korotkov-verkada commented:
Looks like missing resource tracking annotation `app.kubernetes.io/instance` was the reason. I just did a test with adding it to the ConfigMap together with this fix, and there's been a trigger of refresh triggered by config map shortly after it was created. It took less than 30 seconds of waiting for a healthy state of config map.

andrii-korotkov-verkada commented:
I actually was testing adding this annotation before this fix and it didn't work by itself. I don't recall if my test with a fix was on a branch with annotation or not.

andrii-korotkov-verkada commented:
In another test it took ~5 minutes - app refresh based on the config map of interest arrived late for unclear reason.

andrii-korotkov-verkada commented:
Interesting that an update above happened a couple of seconds after a watch for ConfigMap got restarted.

daftping commented:
This is not directly related to this PR, but I want to mention that we observe similar behavior on large clusters where changes to ConfigMaps (or other objects) are significantly delayed. The delay between when Argo CD receives watch events and when it processes them can be as long as 7 minutes or more.

We've found that this issue is particularly pronounced in large clusters with frequently modified objects. Unfortunately, vertically scaling the application controller or tuning its parameters does not help. Splitting one cluster into multiple shards is also not feasible. The event processing logic appears to be a bottleneck.

Here are some metrics showing how Argo CD choked and stopped processing events:
[Conversation in slack](https://cloud-native.slack.com/archives/C04SURUPDL2/p1718308471766089)

I have a reproducible step for this issue, but I haven’t filed the issue yet.

andrii-korotkov-verkada commented:
> This is not directly related to this PR, but I want to mention that we observe similar behavior on large clusters where changes to ConfigMaps (or other objects) are significantly delayed. The delay between when Argo CD receives watch events and when it processes them can be as long as 7 minutes or more.
> 
> We've found that this issue is particularly pronounced in large clusters with frequently modified objects. Unfortunately, vertically scaling the application controller or tuning its parameters does not help. Splitting one cluster into multiple shards is also not feasible. The event processing logic appears to be a bottleneck.
> 
> Here are some metrics showing how Argo CD choked and stopped processing events: [Conversation in slack](https://cloud-native.slack.com/archives/C04SURUPDL2/p1718308471766089)
> 
> I have a reproducible step for this issue, but I haven’t filed the issue yet.

I wonder if some delays in processing are due to the same issue this PR tries to address - race condition and processing app operation on a stale data. If you can and want to, feel free to test this PR in your environment and see if it changes anything. From my observations so far, waiting time for config maps are usually within 1-3 minutes, though sometimes a bit longer like 5-7 minutes. But I don't see 10+ or 20+ minutes which was happening before.

andrii-korotkov-verkada commented:
One more idea I have is to reduce the timeout of watches from 10min to 5min or so. Not entirely sure it'd help as much, but one case of update after watch just restarted hints that it might.

andrii-korotkov-verkada commented:
With watch resync of 5 min most observed waiting finished within 1-3 min, rarely ~5 min. I think it's a good change.

agaudreault commented on controller/appcontroller.go:
```diff
@@ -1480,6 +1478,9 @@ func (ctrl *ApplicationController) processAppRefreshQueueItem() (processNext boo
 			log.Errorf("Recovered from panic: %+v\n%s", r, debug.Stack())
 		}
 		ctrl.appRefreshQueue.Done(appKey)
+		// In most cases we want to have app operation update happen after the sync, so
```
_Commenting on a line so we can have a discussion in a thread._
Can you share metrics

andrii-korotkov-verkada commented:
I've updated the PR to revert a part of changes.

agaudreault commented:
From a code review perspective, the code change seems to make sense. However, since this changes something in the main reconciliation/sync loop, I think it would be important to be able to visualize the effect that this change has. Apart from manual observations, are there any logs/metrics that would allow us to see the improvement and monitor the effect of the change? 

From the comments above, it would seem like there is a race condition, but this does not fix it, it only reduces its occurence. How can we find when the race condition happens? are there any observable data/logs that allows to detect it and how many times it happens?

Without objective data, it is hard to judge if this change improves the sync behavior, without side effects.

andrii-korotkov-verkada commented:
> From a code review perspective, the code change seems to make sense. However, since this changes something in the main reconciliation/sync loop, I think it would be important to be able to visualize the effect that this change has. Apart from manual observations, are there any logs/metrics that would allow us to see the improvement and monitor the effect of the change?
> 
> From the comments above, it would seem like there is a race condition, but this does not fix it, it only reduces its occurence. How can we find when the race condition happens? are there any observable data/logs that allows to detect it and how many times it happens?
> 
> Without objective data, it is hard to judge if this change improves the sync behavior, without side effects.

I've been testing this with logs in Datadog with filters like `service:argocd,container_name:argocd-application-controller,"waiting for healthy state of /ConfigMap",env:staging`. Then I'd look for logs with "Updating operation state" and "Resuming in-progress operation" for some components/applications. They'd roughly show me a timeline of how the status is changing from waiting for many resources to waiting for a config map to then operations running. For example
![Screenshot 2024-07-02 at 8 05 50 AM](https://github.com/argoproj/argo-cd/assets/137232734/b131ccf0-bb28-4209-870c-759feb65c10e)
It won't show cases where wait was almost non-existent, but still reveals max waiting times.

Regarding the race condition, I think most likely this particular instance of it is getting fixed. Since there's refresh and operation queue deduplication for the same app key, that may not be 100%, but also there might not be much we can do. However, extra waiting time for config maps seems to be coming from delay in Kubernetes watch - based on logs of "Requesting app refresh" with structured data corresponding to a config map. Workqueue depth is pretty low, so we might just be in a state of many objects in the cluster with many updates.

andrii-korotkov-verkada commented:
FWIW, the code has been running for us for a few weeks now and I haven't gotten new complaints about applications getting stuck waiting for a config map healthy state.

andrii-korotkov-verkada commented:
Hm, just got a case in prod environment with waiting time almost 30 min, and about 20 min was spent until the signal from Kubernetes watch requested refresh. Let me remove "Fixes" comment from the description. I'm still pretty sure it's a good change, but we need more than this.

andrii-korotkov-verkada commented:
@agaudreault, do you know whether an insertion of app into the refresh queue while the same app's refresh is being processed would not be inserted (until task done is called)? I know that those queues do deduplication, just not 100% sure it happens if the key wasn't retrieved or can also happen if the key was retrieved, but task done wasn't called.

I'm pretty sure that my change is strictly positive (doesn't make sense to do another sync until refresh is done after object was updated, also see more comments above). The remaining problem is not a race condition, but that Kubernetes watch API update of object is processed with a major delay sometimes. This might be due to having lock here https://github.com/argoproj/gitops-engine/blob/a22b34675f57696d55cb9d4abf99f7b07a166654/pkg/cache/cluster.go#L1107-L1126 and a lot of updates in our clusters. But when the update does arrive, if sync operation is executed before the refresh (which might also take some time until https://github.com/argoproj/argo-cd/issues/18929 and https://github.com/argoproj/gitops-engine/issues/600 are done), we'd need to wait for another update of some object.

andrii-korotkov-verkada commented:
Here's a PR to hopefully help with watch events processing https://github.com/argoproj/gitops-engine/pull/603

andrii-korotkov-verkada commented:
Nvm about my comments about the queue, this shouldn't be a problem, I misunderstood how it works:
```
Stingy: a single item will not be processed multiple times concurrently, and if an item is added multiple times before it can be processed, it will only be processed once.
Multiple consumers and producers. In particular, it is allowed for an item to be reenqueued while it is being processed.
```
Looks like if the item is taken out of the queue and being processed, same item can be enqueued again. This is also consistent with some logs I see.

andrii-korotkov-verkada commented:
Seems like this PR in conjunction with https://github.com/argoproj/gitops-engine/pull/603 solve the problem!
https://github.com/argoproj/gitops-engine/pull/603#issuecomment-2212620560

andrii-korotkov-verkada commented:
Here are some perf views of the system collected following https://github.com/argoproj/argo-cd/issues/13534#issuecomment-1562184630.

The build is from master on 2024/07/07 including https://github.com/argoproj/argo-cd/pull/18972, https://github.com/argoproj/argo-cd/pull/18694, https://github.com/argoproj/gitops-engine/pull/601, https://github.com/argoproj/gitops-engine/pull/603.

![Screenshot 2024-07-09 at 3 47 18 PM](https://github.com/argoproj/argo-cd/assets/137232734/3c1c55ba-8135-4273-9fc3-a400a722868c)
![Screenshot 2024-07-09 at 3 48 02 PM](https://github.com/argoproj/argo-cd/assets/137232734/dffc69f2-80be-4bd8-8ebb-dd485bc5ced9)
![Screenshot 2024-07-09 at 3 48 39 PM](https://github.com/argoproj/argo-cd/assets/137232734/da4de4c7-d733-4377-b92d-920f1c8ca684)

andrii-korotkov-verkada commented:
@crenshaw-dev, can you take a look at this one when you have time, please? It's a part of the fix of long waiting time for healthy state of config map when syncing (while config map is created quickly), together with the locking improvement PR.

crenshaw-dev commented:
@andrii-korotkov I won't be able to get to this one soon.

andrii-korotkov-verkada commented:
I've added the e2e test for the scenario and also tested without my change in https://github.com/argoproj/argo-cd/pull/19251. The test passes, I guess I'll do a few more tries without my change trying to make race condition to trigger, and probably add a loop.

andrii-korotkov-verkada commented:
I've tried https://github.com/argoproj/argo-cd/pull/19251 with 10 iterations and still didn't have it timeout. Seems like getting stuck is hard to reproduce. But at least the e2e test covers complete breakage of the flow.

andrii-korotkov-verkada commented:
I've added small optimization for a common case where delay on app update is 0 (default value) - removing extra putting in the app operation queue. I've also re-instanced additional putting to app queue on app addition with the same logic - first refresh and then do sync on fresher data.

andrii-korotkov-verkada commented:
I've also swapped the order of adding to app operation queue and marking the refresh task done.

andrii-korotkov-verkada commented:
Can't repro the issue on master with a simple e2e test even when running 100 times.

andrii-korotkov-verkada commented:
A summary after recent updates:
* I've added a simple e2e test with adding several config maps with different sync waves and tested it separately with current master on 100 iterations https://github.com/argoproj/argo-cd/pull/19251. I was not able to reproduce the sync getting stuck with it. Note, that I've updated the health check to cover the case when health is not present, which would cause a panic before and which seems like a common case for config maps when they got in a good non-missing state.
* In a common case of resource event the amount of items put into the refresh and operation queue remain the same, they are just sequenced now.
* When an app gets added, the amount of items put into the refresh and operation queue remain the same, they are just sequenced now.
* When an app gets updated, the amount of items put into the refresh queue remains the same, while for operation queue it would be the same or less (if delay is 0, we don't put extra item).
* In a rare case of an app getting deleted, there'd be one extra put in the operation queue, but it should be processed quickly due to a check that app exists.
@alexmt, please take a look when you have time and see if concerns were addressed well.

alexmt reviewed (approved):
LGTM.

Notes from our discussion. Adding app to both queues at the same time for a mistake, since definitely don't want to reconcile sync status and operation at the same time.


I'm not concerned about performance as well. The `requestAppRefresh` is executed very frequently e.g. when child resource like Pod or ReplicaSet changes status. So it will reduce number of app operation reconciliations.
